Lib/asyncio/tasks.py

```diff
@@ -439,63 +440,23 @@ async def wait_for(fut, timeout):
 
     This function is a coroutine.
     """
-    loop = events.get_running_loop()
-
     if timeout is None:
         return await fut
 
     if timeout <= 0:
-        fut = ensure_future(fut, loop=loop)
+        fut = ensure_future(fut)
 
         if fut.done():
             return fut.result()
 
-        await _cancel_and_wait(fut, loop=loop)
+        await _cancel_and_wait(fut)
         try:
             return fut.result()
         except exceptions.CancelledError as exc:
-            raise exceptions.TimeoutError() from exc
-
-    waiter = loop.create_future()
-    timeout_handle = loop.call_later(timeout, _release_waiter, waiter)
-    cb = functools.partial(_release_waiter, waiter)
-
-    fut = ensure_future(fut, loop=loop)
-    fut.add_done_callback(cb)
-
-    try:
-        # wait until the future completes or the timeout
-        try:
-            await waiter
-        except exceptions.CancelledError:
-            if fut.done():
-                return fut.result()
-            else:
-                fut.remove_done_callback(cb)
-                # We must ensure that the task is not running
-                # after wait_for() returns.
-                # See https://bugs.python.org/issue32751
-                await _cancel_and_wait(fut, loop=loop)
-                raise
-
-        if fut.done():
-            return fut.result()
-        else:
-            fut.remove_done_callback(cb)
-            # We must ensure that the task is not running
-            # after wait_for() returns.
-            # See https://bugs.python.org/issue32751
-            await _cancel_and_wait(fut, loop=loop)
-            # In case task cancellation failed with some
-            # exception, we should re-raise it
-            # See https://bugs.python.org/issue40607
-            try:
-                return fut.result()
-            except exceptions.CancelledError as exc:
-                raise exceptions.TimeoutError() from exc
-    finally:
-        timeout_handle.cancel()
+            raise TimeoutError from exc
```

Comment:
Somewhere here (or in `_cancel_and_wait()`) we ought to call `uncancel()`, but only if _we_ are the party that called `cancel()`, and then only if an `uncancel` method actually exists. (It may not if it's a future, or if it's a task created by a 3rd party event loop that doesn't yet support `ucancel()`.)

Reply:
Why is that required? I haven't used `uncancel` much so might be missing the context.

Reply:
Uncancel is the great mystery of asyncio. :-)

Seriously, when things like wait_for() or timeout() are nested, it is possible that two timeouts occur "simultaneously", and a task gets cancelled twice before its coroutine gets the CancelledError thrown in. It's no problem for the task (it should just clean up and reraise), but the inner timeout/wait_for call should _not_ raise a TimeoutError, but instead bubble up the CancelledError.

Why not raise TimeoutError? Because the outer thing might not be a timeout -- it might be a TaskGroup or some other thing. For example, we could have something like "happy eyeballs" that wants to cancel all other attempts once the first attempt has a result. But it might also set timeouts on other attempts. The cancellation should preempt the timeout.

To implement such cooperative cancellation schemes, Task.cancel() increments a counter, and code that cancels a task and later handles the cancellation is supposed to call uncancel(), which decrements the counter and returns the remaining value. If that's nonzero, there was another cancel that had priority (because its handling will happen in an outer construct).

The timeout() context manager (and TaskGroup() too) implements this collaborative scheme, I think wait_for() should too.

Reply:
Thanks for the explanation though I feel it would be better if you do this as you understand `uncancel`  better than me.
This PR only targets 3.12 so we have time to do this, so leaving this for another day (for now).